### PR TITLE
Enable reporting of undefined symbols in wasm-ld

### PR DESCRIPTION
**Pull request by sbc100**
When enabled, wasm-ld will report more detailed information about the
undefined symbols such as which object file/funcion required the
symbols.

This works by generating the list of all possible JS-defined symbols
and passing that list of wasm-ld so it can determine which symbols
are permitted to be undefined.

The cost of this feature is that we run the js compiler twice, once
before wasm-ld and once after the link once wasm-emscripten-finalize
has been run.  On my machine this takes around 700 so I'm not enabling
this by default yet.

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -512,6 +512,38 @@ def ensure_archive_index(archive_file):
     run_process([shared.LLVM_RANLIB, archive_file])
 
 
+def get_all_js_library_funcs(temp_files):
+  # Runs the js compiler to generate a list of all functions available in the JS
+  # libraries.  This must be done sepraetly for each linker invokation since the
+  # list of library functions depends on what settings are used.
+  # TODO(sbc): Find a way to optmize this.  Potenally we could add a super-set
```
Comment (kripken) [inline emcc.py]: typo in `potentially`

File: emcc.py [hunk 2]
```diff
@@ -512,6 +512,38 @@ def ensure_archive_index(archive_file):
...
+    emscripten.generate_struct_info()
+    glue, forwarded_data = emscripten.compile_settings(temp_files)
+    forwarded_json = json.loads(forwarded_data)
+    library_fns = forwarded_json['Functions']['libraryFunctions']
+    library_fns_list = []
+    for name in library_fns:
+      if name[0] == '_':
+        name = name[1:]
+        library_fns_list.append(name)
+        library_fns_list.append('__wasi_' + name)
```
Comment (kripken) [inline emcc.py]: should this use `WASI_MODULE_NAME`?

File: emcc.py [hunk 3]
```diff
@@ -1146,6 +1178,7 @@ def check(input_file):
 
     if shared.Settings.STRICT:
       shared.Settings.STRICT_JS = 1
+      shared.Settings.LLD_REPORT_UNDEFINED = 1
```
Comment (kripken) [inline emcc.py]: why enable this in strict? I'm not happy with strict compiles being slower.

if we have an error, can we show "link with -s LLD_REPORT_UNDEFINED to get more detailed error reporting"?

File: emscripten.py [hunk 4]
```diff
@@ -2769,11 +2769,8 @@ def normalize_line_endings(text):
   return text
 
 
-def run(infile, outfile, memfile):
-  temp_files = get_configuration().get_temp_files()
-  infile, outfile = substitute_response_files([infile, outfile])
-
-  if not shared.Settings.BOOTSTRAPPING_STRUCT_INFO:
+def generate_struct_info():
+  if not shared.Settings.STRUCT_INFO and not shared.Settings.BOOTSTRAPPING_STRUCT_INFO:
```
Comment (kripken) [inline emscripten.py]: why has STRUCT_INFO been added as a check here?

File: src/library.js [hunk 5]
```diff
@@ -4819,12 +4819,12 @@ LibraryManager.library = {
...
-  emscripten_asm_const_double: true,
-  emscripten_asm_const_int_sync_on_main_thread: true,
-  emscripten_asm_const_double_sync_on_main_thread: true,
-  emscripten_asm_const_async_on_main_thread: true,
+  emscripten_asm_const: function() {},
+  emscripten_asm_const_int: function() {},
+  emscripten_asm_const_double: function() {},
+  emscripten_asm_const_int_sync_on_main_thread: function() {},
+  emscripten_asm_const_double_sync_on_main_thread: function() {},
+  emscripten_asm_const_async_on_main_thread: function() {},
```
Comment (kripken) [inline src/library.js]: why is this needed, and why just these functions?

File: emcc.py (hunk 3 above)
Comment (sbc100) [reply emcc.py]: Good idea.  Done.

File: src/library.js (hunk 5 above)
Comment (sbc100) [reply src/library.js]: Without this change the list of functions in `forwarded_json['Functions']['libraryFunctions']` won't include these since (I suppose) it sees them as data rather than functions?

We need the list of include these functions otherwise this new flag won't work with any EM_ASM-using code.

The comment above here says `misc definitions to avoid unnecessary unresolved symbols from fastcomp`.  Which I guess means that fastcomp undefined symbol checker also depends on these, but that is does't care if they are functions of non-functions.   I will update the command and apply this change to all those other ones too.

File: emscripten.py (hunk 4 above)
Comment (sbc100) [reply emscripten.py]: This avoids generating the struct into twice.

In the old code we would do:

```
if not BOOTSTRAPPING_STRUCT_INFO:
   STRUCT_INFO = ..do work ..;
```

However in the new layout we might end up calling this code twice so I added the extra guard here which basically says "if we have already got the struct info from the cache then don't bother getting it again".  Perhaps we should call this function `load_struct_info` since it doesn't always generate it.

File: emcc.py (hunk 2 above)
Comment (sbc100) [reply emcc.py]: This isn't actually WASI_MODULE_NAME though.. its something a little different.  I'll add a comment.

**Comment by sbc100** [comment]
Should be good to go now.

File: ChangeLog.md [hunk 6]
```diff
@@ -17,6 +17,10 @@ See docs/process.md for how version tagging works.
 
 Current Trunk
 -------------
+- Add LLD_REPORT_UNDEFINED option that should allow for more details
```
Comment (kripken) [inline ChangeLog.md]: typo in `detailed`
